sync/server.md

```diff
@@ -0,0 +1,404 @@
+The result of the SQL query will be made available on the `context.result` object.
+The `result` field will always be an `array`.
+If a list of objects is expected in the Schema rsponse, this can be mapped as follows:
+
+```
+{{toJSON context.result}}
+```
+
+If a max of 1 result is expected, e.g. finding a record by id, this can be mapped as follows:
+
+```
+{{toJSON context.result.[0]}}
+```
+
+## HTTP Connector
+
+### Data Source
+
+The defintion of a HTTP Data Source will have the following parameters:
+
+* Protocol i.e. HTTP or HTTPS
+* Host e.g. www.example.com
+* Port
+* Base path e.g. /myapi/v1
+* Headers (optional)
+* Basic Auth Username (optional)
+* Basic Auth Password (optional)
+
+### Resolver Request Mapping
+
+This is the defintion of a HTTP request to make for a specific Query or Mutation.
+It will have the following parameters in conjunciton with the Data Source parameters to make a request.
+
+* Path (Added to the end of the Base path defined in the Data Source. Can include query parameters)
+* Additional Headers (optional)
+* Method e.g. GET, POST, PUT
+* Request Body (Optional)
+
+These will be defined as a JSON object, but will be treated as a string.
+This means the below defintion would actually be stored as the string beneath
+
+JSON version
+```
+{
+  "path": "/notes",
+  "method": "GET",
+  "headers": {
+    "accept": "application/json"
+  }
+}
+```
+
+String version
+```
+{\n  \"path\": \"/notes\",\n  \"method\": \"GET\",\n  \"headers\": {\n    \"accept\": \"application/json\"\n  }}
+```
+
+By treating the defintion as a string, it allows for Handlebars syntax to be added anywhere in the JSON object.
+For example, the below JSON defintion could be used for a simple find Query.
+
+```
+{
+  "path": "/notes?id={{context.arguments.id}}",
+  "method": "GET",
+  "headers": {
+    "accept": "application/json"
+  }
+}
+```
+
+### Resolver Response Mapping
+
+This is the defintion for how to format the result of the HTTP Request.
+The response body will be made available on `context.result`.
+It's type depends on the `Accept` header in the request.
+Initially, only `application/json` will have special behaviour, meaning the response body is parsed into a JSON object.
+All other `Accept` header values will be ignored, in which case `context.result` will be the plain response body as a string.
+
+If the result from the HTTP call is exactly as you need it, you can simply dump out the result.
+
+```
+{{toJSON context.result}}
+```
+
+If you need to iterate through the response as a JSON array, mapping fields, you can do something like below.
+In this example, each item has a number of fields, one of which is `name`.
+However, this needs to be changed to `firstname` to match our schema.
+
+```
+TODO: Helper for mapping an array
+TODO: Is this a reasonable use case for response mapping? Maybe filtering is better e.g. omitting records that don't have a paritcular field set
+
+{{toJSON mappedResult}}
+```
```

Comment:
@david-martin I think the default request and response resolvers are great for certain use cases, but I am not sure they are flexible enough to support some real-world use cases.

For example, in my resolver, I want to first fetch data from an external source via http. But the external source could be unstable, so I would like to cache/persist the data locally and if the external source is not available, the cached version of the data will be returned.

Is it possible to support this scenario in the resolvers?

I am also thinking can we allow developers to actually define the resolvers in nodejs? We can take advantage of the [VM module](https://nodejs.org/api/vm.html) to run the scripts in a sandboxed environment, but it will allow developers to do more if their use case is a bit more complicated. 

Do you think it is something worth spiking?

Reply:
For custom code, there are 2 options:

* an openwhisk data source with custom code defined as actions
* write a custom server that exposes a http endpoint, and use the http data source